### ISSUE TITLE
[feature][sponsored] package cleanup before  

### ISSUE THREAD

**Issue by GeoffWilliams** [albatrossflavour/puppet_os_patching#55]
Sometimes you will want do a package cleanup:
* `yum clean all`
* `apt-get clean`

Before running the main update. Support this via a new task argument `clean_pkg_cache`

**Comment by TJM** [comment]
In addition to this, it might be nice to have the "package_cache_cleanup'' ("yum clean all" or "apt-get clean") as a task. When we update our pulp views, we sometimes have to run "yum clean all" before the updates will show up in the "yum check-updates" (aka os_patching fact)

**Comment by GeoffWilliams** [comment]
@TJM So do you think the yum/apt/??? pkg clean should be its own step (as well?instead?) we're still bouncing idea around as to how best service this at the moment

**Comment by TJM** [comment]
@GeoffWilliams Not sure whether it should be "as well" or "instead" ... I was thinking "instead" (as in its own job) because I want the os_patching fact to be correct, since it is part of the selection criteria when running the patching task. If it is not showing any updates *because* we need to run "yum clean all" then running the yum clean all before patching won't do any good for us because we won't select it for running the patch job.

I was actually wondering whether it shouldn't just be part of the os_patching_fact_generation.sh process? I suppose I could control that with the yumrepo settings too, similar to how the RHN channels force update for every yum run. :-/

I kinda wish yum was smart about it and would do like a "HEAD" request to see if it needed to update the file :)

**Comment by albatrossflavour** [comment]
Putting it into the fact generation will put a fair bit more load on the 
yum servers if we do it each time.

I like the idea of having it in the task as a parameter, but that can 
mask duff update data if the caches are incorrect.

Maybe we attack from 3 sides.

* Have the parameter to the task to allow a clean prior to running
* Have a dedicated task to do a clean that is manually triggered
* Update the fact generation script to check if the `cachecookie` is 
over XX hours old (which could be a parameter to the fact script) and, 
if so, trigger a clean

Thoughts?

**Comment by GeoffWilliams** [comment]
I like this approach

**Comment by albatrossflavour** [comment]
https://github.com/albatrossflavour/puppet_os_patching/tree/feature/clean_cache

I'm working on it

**Comment by GeoffWilliams** [comment]
Not an expert by any means but what I've seen with customers managing their own repositories (eg artifactory) is that sometimes for whatever reason, the metadata cache needs to be invalidated fleet-wide because of meddling in the corporate repository - eg, replace say java 8 with java 7 and then install "java" everywhere. Of course this isn't necessarily the "right" way to do things but this is what happens in the real world. For doing stuff in facter, I'm not too vested in the whole cache cookie thing - if we can just do a yum clean somwhere, that would keep my customer happy. @TJM Is this the kind of thing your seeing too?

**Comment by albatrossflavour** [comment]
I totally agree, just saying not to keen on making the module a swiss army knife of fixing issues that should be fixed elsewhere.  I've got the task param for patching working and the dedicated clean cache task will be there shortly.  

The effort and accuracy to 'guess' if we should do the clean as part of the facter cache generation is all I want to pull.

**Comment by GeoffWilliams** [comment]
ah.. now I get you. Maybe its time to make the sysadmins earn their keep:
1. parameter on main update task to `yum clean` as part of update
2. separate task to run `yum clean`
3. if yum doesn't figure out it's metadata is out of date (eg someone was meddling with the repo) - tough - let it fail, don't try to second-guess it with factor etc. The human operator can figure this out and either run the clean task or update with clean parameter and that's it.

Is that what your proposing? Sounds good to me

**Comment by GeoffWilliams** [comment]
also.. did I read "do a clean as part of a factor run"? Facter should not be modifying the system at all... look with your eyes not your hands ;-)

**Comment by albatrossflavour** [comment]
Yep, 100% what I'm saying.  Otherwise we'll get into the area of trying to diagnose and report on other borkages within yum etc.

**Comment by GeoffWilliams** [comment]
lets do that - sounds good to me

**Comment by albatrossflavour** [comment]
New task to do cache cleans and a parameter to the patching task to do the same have been added as part of #75

**Comment by TJM** [comment]
What I generally see is... Our customers mirror outside repos either with pulp or "mrepo" or something like that to an internal server. They update their test/dev environment first, then after a week or so, the staging/qa environment, then after another week, production. They want the ability to *validate* the patches before applying them to production. After patches are applied to production, the pulp/mrepo process starts over and they sync with the outside repos again. It is after that step that someone might want to run `yum clean all`, because you may want to start testing the patches immediately on some test/dev systems.

Basically the cache is working the way it is designed to, but some "event" (update server sync'd) makes it so that you want to invalidate the cache.

I am totally happy with how this came out. Sorry, it appears we are on opposite sides of the world, so you guys already solved this while I was sleeping :)

~tommy
